File: backend-java/src/main/java/com/peopleminer/scoring/ScoringService.java
package com.peopleminer.scoring;

import com.peopleminer.domain.entity.Candidate;
import com.peopleminer.domain.repository.CandidateRepository;
import com.peopleminer.scoring.StrategyRegistry.StrategyWithConfig;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class ScoringService {

    private final CandidateRepository candidateRepository;
    private final StrategyRegistry strategyRegistry;

    @Transactional
    public ScoringResult scoreCandidate(String candidateId) {
        Candidate candidate = candidateRepository.findByIdWithRelations(candidateId)
            .orElseThrow(() -> new IllegalArgumentException("Candidate " + candidateId + " not found"));

        return calculateScore(candidate);
    }

    @Transactional
    public ScoringResult calculateScore(Candidate candidate) {
        List<StrategyWithConfig> enabledStrategies = strategyRegistry.getEnabledStrategies();
        Map<String, Double> normalizedWeights = strategyRegistry.getNormalizedWeights();

        List<StrategyScoreResult> strategyScores = new ArrayList<>();
        double totalScore = 0;

        for (StrategyWithConfig swc : enabledStrategies) {
            try {
                ScoringStrategy strategy = swc.getStrategy();

                if (strategy.isApplicable(candidate)) {
                    ScoringStrategy.StrategyScore result = strategy.calculate(candidate);
                    double normalizedWeight = normalizedWeights.getOrDefault(swc.getConfig().getName(), 0.0);
                    double weightedScore = result.getValue() * normalizedWeight;

                    strategyScores.add(StrategyScoreResult.builder()
                        .strategyName(strategy.getName())
                        .score(result.getValue())
                        .weight(normalizedWeight)
                        .weightedScore(weightedScore)
                        .breakdown(result.getBreakdown())
                        .build());

                    totalScore += weightedScore;
                }
            } catch (Exception e) {
                log.error("Strategy {} failed for candidate {}: {}",
                    swc.getStrategy().getName(), candidate.getId(), e.getMessage());
            }
        }

        totalScore = Math.round(totalScore * 100) / 100.0;

        // Update candidate scores
        candidate.setTotalScore(totalScore);
        candidate.setReadabilityScore(findStrategyScore(strategyScores, "codeQuality"));
        candidate.setProblemSolvingScore(findStrategyScore(strategyScores, "problemSolving"));
        candidate.setCleanCodeScore(findStrategyScore(strategyScores, "activity"));
        candidate.setSolvedAcScore(findStrategyScore(strategyScores, "solvedAc"));
        candidate.setScoredAt(LocalDateTime.now());

        candidateRepository.save(candidate);

        return ScoringResult.builder()
            .candidateId(candidate.getId())
            .totalScore(totalScore)
            .strategyScores(strategyScores)
            .scoredAt(LocalDateTime.now())
            .build();
    }

    private Double findStrategyScore(List<StrategyScoreResult> scores, String strategyName) {
        return scores.stream()
            .filter(s -> s.getStrategyName().equals(strategyName))
            .map(StrategyScoreResult::getScore)
            .findFirst()
            .orElse(null);
    }

    @Transactional
    public ScoreBatchResult scoreAllCandidates(boolean force, int batchSize) {
        LocalDateTime cutoff = LocalDateTime.now().minusDays(7);
        List<Candidate> candidates = force
            ? candidateRepository.findAll(PageRequest.of(0, batchSize)).getContent()
            : candidateRepository.findCandidatesNeedingScoring(cutoff, PageRequest.of(0, batchSize));

        int scored = 0;
        int failed = 0;

        for (Candidate candidate : candidates) {
            try {
                calculateScore(candidate);
                scored++;

                if (scored % 10 == 0) {
                    log.info("Scored {}/{} candidates", scored, candidates.size());
                }
            } catch (Exception e) {
                log.error("Failed to score candidate {}: {}", candidate.getId(), e.getMessage());
                failed++;
            }
        }

        log.info("Scoring complete: {} scored, {} failed", scored, failed);
        return ScoreBatchResult.builder()
            .scored(scored)
            .failed(failed)
            .build();
    }

    public List<StrategyInfo> getStrategies() {
        return strategyRegistry.getAllStrategies().stream()
            .map(swc -> StrategyInfo.builder()
                .name(swc.getStrategy().getName())
                .description(swc.getStrategy().getDescription())
                .enabled(swc.getConfig().getEnabled())
                .weight(swc.getConfig().getWeight())
                .defaultWeight(swc.getStrategy().getDefaultWeight())
                .build())
            .toList();
    }

    public void updateStrategyWeight(String name, double weight) {
        strategyRegistry.setWeight(name, weight);
    }

    public void enableStrategy(String name) {
        strategyRegistry.enable(name);
    }

    public void disableStrategy(String name) {
        strategyRegistry.disable(name);
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class ScoringResult {
        private String candidateId;
        private double totalScore;
        private List<StrategyScoreResult> strategyScores;
        private LocalDateTime scoredAt;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class StrategyScoreResult {
        private String strategyName;
        private double score;
        private double weight;
        private double weightedScore;
        private Map<String, Double> breakdown;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class ScoreBatchResult {
        private int scored;
        private int failed;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class StrategyInfo {
        private String name;
        private String description;
        private Boolean enabled;
        private Double weight;
        private double defaultWeight;
    }
}

File: backend/src/scoring/scoring.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import {
  StrategyRegistry,
  ScoringResult,
  CandidateWithRelations,
} from './strategies';

@Injectable()
export class ScoringService {
  private readonly logger = new Logger(ScoringService.name);

  constructor(
    private prisma: PrismaService,
    private strategyRegistry: StrategyRegistry,
  ) {}

  async scoreCandidate(candidateId: string): Promise<ScoringResult> {
    const candidate = await this.prisma.candidate.findUnique({
      where: { id: candidateId },
      include: {
        repositories: {
          include: { analysis: true },
        },
        sources: true,
        solvedAcProfile: true,
        ossContributions: true,
      },
    });

    if (!candidate) {
      throw new Error(`Candidate ${candidateId} not found`);
    }

    return this.calculateScore(candidate);
  }

  async calculateScore(candidate: CandidateWithRelations): Promise<ScoringResult> {
    const enabledStrategies = this.strategyRegistry.getEnabledStrategies();
    const normalizedWeights = this.strategyRegistry.getNormalizedWeights();

    const strategyScores: ScoringResult['strategyScores'] = [];
    let totalScore = 0;

    for (const { strategy, config } of enabledStrategies) {
      try {
        if (strategy.isApplicable && !strategy.isApplicable(candidate)) {
          continue;
        }

        const result = await strategy.calculate(candidate);
        const normalizedWeight = normalizedWeights.get(config.name) || 0;
        const weightedScore = result.value * normalizedWeight;

        strategyScores.push({
          strategyName: strategy.name,
          score: result.value,
          weight: normalizedWeight,
          weightedScore,
          breakdown: result.breakdown,
        });

        totalScore += weightedScore;
      } catch (error) {
        this.logger.error(`Strategy ${strategy.name} failed for ${candidate.id}:`, error);
      }
    }

    totalScore = Math.round(totalScore * 100) / 100;

    await this.prisma.candidate.update({
      where: { id: candidate.id },
      data: {
        totalScore,
        readabilityScore: this.findStrategyScore(strategyScores, 'codeQuality'),
        problemSolvingScore: this.findStrategyScore(strategyScores, 'problemSolving'),
        cleanCodeScore: this.findStrategyScore(strategyScores, 'activity'),
        solvedAcScore: this.findStrategyScore(strategyScores, 'solvedAc'),
        scoredAt: new Date(),
      },
    });

    return {
      candidateId: candidate.id,
      totalScore,
      strategyScores,
      scoredAt: new Date(),
    };
  }

  private findStrategyScore(
    scores: ScoringResult['strategyScores'],
    strategyName: string,
  ): number | null {
    const found = scores.find(s => s.strategyName === strategyName);
    return found?.score ?? null;
  }

  async scoreAllCandidates(options?: {
    force?: boolean;
    batchSize?: number;
  }): Promise<{ scored: number; failed: number }> {
    const { force = false, batchSize = 50 } = options || {};

    const whereClause = force
      ? {}
      : {
          OR: [
            { scoredAt: null },
            { scoredAt: { lt: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000) } },
          ],
        };

    const candidates = await this.prisma.candidate.findMany({
      where: whereClause,
      include: {
        repositories: {
          include: { analysis: true },
        },
        sources: true,
        solvedAcProfile: true,
        ossContributions: true,
      },
      take: batchSize,
    });

    let scored = 0;
    let failed = 0;

    for (const candidate of candidates) {
      try {
        await this.calculateScore(candidate);
        scored++;

        if (scored % 10 === 0) {
          this.logger.log(`Scored ${scored}/${candidates.length} candidates`);
        }
      } catch (error) {
        this.logger.error(`Failed to score candidate ${candidate.id}:`, error);
        failed++;
      }
    }

    this.logger.log(`Scoring complete: ${scored} scored, ${failed} failed`);
    return { scored, failed };
  }

  getStrategies() {
    return this.strategyRegistry.getAllStrategies().map(({ strategy, config }) => ({
      name: strategy.name,
      description: strategy.description,
      enabled: config.enabled,
      weight: config.weight,
      defaultWeight: strategy.defaultWeight,
    }));
  }

  updateStrategyWeight(name: string, weight: number) {
    this.strategyRegistry.setWeight(name, weight);
  }

  enableStrategy(name: string) {
    this.strategyRegistry.enable(name);
  }

  disableStrategy(name: string) {
    this.strategyRegistry.disable(name);
  }
}
